src/Orleans.Transactions/State/ConfirmationWorker.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Orleans.Configuration;
+using Orleans.Transactions.Abstractions;
+
+namespace Orleans.Transactions.State
+{
+    internal class ConfirmationWorker<TState>
+        where TState : class, new()
+    {
+        private readonly TransactionalStateOptions options;
+        private readonly ParticipantId me;
+        private readonly BatchWorker storageWorker;
+        private readonly Func<StorageBatch<TState>> getStorageBatch;
+        private readonly ILogger logger;
+        private readonly HashSet<Guid> pending;
+
+        public ConfirmationWorker(IOptions<TransactionalStateOptions> options, ParticipantId me, BatchWorker storageWorker, Func<StorageBatch<TState>> getStorageBatch, ILogger logger)
+        {
+            this.options = options.Value;
+            this.me = me;
+            this.storageWorker = storageWorker;
+            this.getStorageBatch = getStorageBatch;
+            this.logger = logger;
+            this.pending = new HashSet<Guid>();
+        }
+
+        public void Add(Guid transactionId, DateTime timestamp, List<ParticipantId> participants)
+        {
+            if(!IsConfirmed(transactionId))
+            {
+                this.pending.Add(transactionId);
+                SendConfirmation(transactionId, timestamp, participants).Ignore();
+            }
+        }
+
+        public bool IsConfirmed(Guid transactionId)
+        {
+            return this.pending.Contains(transactionId);
+        }
+
+        private async Task SendConfirmation(Guid transactionId, DateTime timestamp, List<ParticipantId> participants)
+        {
+            await NotifyAll(transactionId, timestamp, participants);
+            await Collect(transactionId);
+        }
+
+        private async Task NotifyAll(Guid transactionId, DateTime timestamp, List<ParticipantId> participants)
+        {
+            List<Confirmation> confirmations = participants
+                    .Where(p => !p.Equals(this.me))
+                    .Select(p => new Confirmation(
+                        p,
+                        transactionId,
+                        timestamp,
+                        () => p.Reference.AsReference<ITransactionalResourceExtension>()
+                            .Confirm(p.Name, transactionId, timestamp),
+                        this.logger))
+                    .ToList();
+
+            // attempts to confirm all, will retry every ConfirmationRetryDelay until all succeed
+            while ((await Task.WhenAll(confirmations.Select(c => c.Confirmed()))).Any(b => !b))
```

Comment:
is it ok to just keep retry without a limit? what if one of the participant is stuck and never return ? what's the expected behavior in this case?

Reply:
Unfortunately algorithm requires confirmations eventually succeed.  This logic only retries those that failed, but it will retry forever (or until grain/silo go away).

Reply:
Side note, original logic did this as well, just in a different manner.